xray/segment.go

```diff
@@ -286,34 +296,56 @@ func (seg *Segment) RemoveSubsegment(remove *Segment) bool {
 			seg.rawSubsegments = seg.rawSubsegments[:len(seg.rawSubsegments)-1]
 
 			if seg.ParentSegment != seg {
-				seg.ParentSegment.Lock()
-				defer seg.ParentSegment.Unlock()
+				seg.openSegments--
+				seg.Unlock()
+				unlock = false
+
+				atomic.AddUint32(&seg.ParentSegment.totalSubSegments, ^uint32(0))
+			}
+			if unlock {
+				seg.Unlock()
 			}
-			seg.ParentSegment.totalSubSegments--
-			seg.openSegments--
 			return true
 		}
 	}
+	seg.Unlock()
 	return false
 }
 
+func (seg *Segment) isOrphan() bool {
+	return seg.parent == nil || seg.Type == "subsegment"
+}
+
 func (seg *Segment) emit() {
 	seg.ParentSegment.GetConfiguration().Emitter.Emit(seg)
 }
 
-func (seg *Segment) handleContextDone() {
+func (seg *Segment)    handleContextDone() {
```

Comment:
Could we move the extra indent before the method?

Reply:
done